File: technical_test.py
from datetime import date


class Person:
    def __init__(self, born_date=None, genre=False, edad=0):
        self.born_date = born_date
        self.genre = genre
        self.edad = edad


class Empleado:
    def __init__(
        self, id=0, nombre="", cargo="", salario=0, fecha=None, edad=0, sexo=""
    ):
        self.id = id
        self.nombre = nombre
        self.cargo = cargo
        self.salario = salario
        self.fecha = fecha
        self.edad = edad
        self.sexo = sexo


class Departamento_comercial(Empleado):
    def __init__(self, venta, ciudad):
        super().__init__()
        self.venta = venta
        self.ciudad = ciudad


class Gerencia(Empleado):
    def __init__(self, pais):
        super().__init__()
        self.pais = pais


import datetime


class Prueba_tecnica:
    def Prueba_tecnica(self):
        id = 97091918500
        fecha_nacimiento = self.obtener_fecha_de_nacimiento(id)
        if date is None:
            return
        edad = self.obtener_edad(fecha_nacimiento)
        sexo = "Masculino" if self.is_man_or_not(id) else "Femenino"

        print("La fecha es:", fecha_nacimiento)
        print("La edad es:", edad)
        print("El sexo es:", sexo)

    def get_empleado_test(self):
        empleado = self.obtener_empleado()
        print("Empleado:", empleado)

    def obtener_empleado(self):
        id = 97091918500
        date = self.obtener_fecha_de_nacimiento(id)
        if date is None:
            return None
        edad = self.obtener_edad(date)
        sexo = "Masculino" if self.is_man_or_not(id) else "Femenino"

        return {
            "id": 34547624768,
            "nombre": "Susan",
            "cargo": "Jefa de Ventas",
            "salario": 3000,
            "fecha": date,
            "edad": edad,
            "sexo": sexo,
        }

    def obtener_salario_total(self, salario, bonos=0):
        return salario + bonos

    def obtener_edad(self, fecha_nacimiento):
        return date.today().year - fecha_nacimiento.year

    def is_man_or_not(self, id):
        digitTen = int(str(id)[10])
        return digitTen % 2 == 1

    def obtener_fecha_de_nacimiento(self, id):
        if len(str(id)) != 11:
            return self.mensaje_de_error()

        carnet = str(id)

        year = int(carnet[0:2])
        month = int(carnet[2:4])
        day = int(carnet[4:6])

        if month > 12 or month < 0:
            return self.mensaje_de_error()
        if day < 0 or day > 31:
            return self.mensaje_de_error()

        if year <= 99:
            year = 1900 + year

        return datetime.date(year, month, day)

    def mensaje_de_error(self):
        print("Error: Número de identidad incorrecto")
        return None


test = Prueba_tecnica()
test.Prueba_tecnica()
test.get_empleado_test()

File: technicalTest.js
class Person {
  constructor(born_date = null, genre = false, edad = 0) {
    this.born_date = born_date;
    this.genre = genre;
    this.edad = edad;
  }
}

class Empleado {
  constructor(
    id = 0,
    nombre = "",
    cargo = "",
    salario = 0,
    fecha = null,
    edad = 0,
    sexo = ""
  ) {
    this.id = id;
    this.nombre = nombre;
    this.cargo = cargo;
    this.salario = salario;
    this.fecha = fecha;
    this.edad = edad;
    this.sexo = sexo;
  }
}

class Departamento_comercial extends Empleado {
  constructor(venta, ciudad) {
    super();
    this.venta = venta;
    this.ciudad = ciudad;
  }
}

class Gerencia extends Empleado {
  constructor(pais) {
    super();
    this.pais = pais;
  }
}

class Prueba_tecnica {
  Prueba_tecnica() {
    const id = 97091918500;
    const fecha_nacimiento = this.obtener_fecha_de_nacimiento(id);
    if (fecha_nacimiento === null) {
      return;
    }
    const edad = this.obtener_edad(fecha_nacimiento);
    const sexo = this.is_man_or_not(id) ? "Masculino" : "Femenino";

    console.log("La fecha es:", fecha_nacimiento.toLocaleDateString());
    console.log("La edad es:", edad);
    console.log("El sexo es:", sexo);
  }

  get_empleado_test() {
    const empleado = this.obtener_empleado();
    console.log("Empleado:", empleado);
  }

  obtener_empleado() {
    const id = 97091918500;
    const date = this.obtener_fecha_de_nacimiento(id);
    if (date === null) {
      return null;
    }
    const edad = this.obtener_edad(date);
    const sexo = this.is_man_or_not(id) ? "Masculino" : "Femenino";

    return {
      id: 34547624768,
      nombre: "Susan",
      cargo: "Jefa de Ventas",
      salario: 3000,
      fecha: date,
      edad: edad,
      sexo: sexo,
    };
  }

  obtener_salario_total(salario, bonos = 0) {
    return salario + bonos;
  }

  obtener_edad(fecha_nacimiento) {
    const today = new Date();
    return today.getFullYear() - fecha_nacimiento.getFullYear();
  }

  is_man_or_not(id) {
    const digitTen = parseInt(id.toString().charAt(10));
    return digitTen % 2 === 1;
  }

  obtener_fecha_de_nacimiento(id) {
    if (id.toString().length !== 11) {
      return this.mensaje_de_error();
    }

    const carnet = id.toString();

    const year = parseInt(carnet.substring(0, 2));
    const month = parseInt(carnet.substring(2, 4));
    const day = parseInt(carnet.substring(4, 6));

    if (month > 12 || month < 0) {
      return this.mensaje_de_error();
    }
    if (day < 0 || day > 31) {
      return this.mensaje_de_error();
    }

    let fullYear = year;
    if (year <= 99) {
      fullYear = 1900 + year;
    }

    return new Date(fullYear, month - 1, day);
  }

  mensaje_de_error() {
    console.error("Número de identidad incorrecto");
    return null;
  }
}

const test = new Prueba_tecnica();
test.Prueba_tecnica();
test.get_empleado_test();
